pkg/sharky/store.go

```diff
@@ -0,0 +1,198 @@
+		pool:     pool,
+		writes:   make(chan write),
+		shards:   make([]*shard, shardCnt),
+		wg:       &sync.WaitGroup{},
+		quit:     make(chan struct{}),
+		metrics:  newMetrics(),
+	}
+	for i := range store.shards {
+		s, err := store.create(uint8(i), datasize, basedir)
+		if err != nil {
+			return nil, err
+		}
+		store.shards[i] = s
+	}
+	store.metrics.ShardCount.Set(float64(len(store.shards)))
+
+	return store, nil
+}
+
+// Close closes each shard and return incidental errors from each shard
+func (s *Store) Close() (err error) {
+	close(s.quit)
+	for _, sh := range s.shards {
+		err = multierror.Append(err, sh.close())
+	}
+
+	return err.(*multierror.Error).ErrorOrNil()
+}
+
+// create creates a new shard with index, max capacity limit, file within base directory
+func (s *Store) create(index uint8, datasize int, basedir fs.FS) (*shard, error) {
+	file, err := basedir.Open(fmt.Sprintf("shard_%03d", index))
+	if err != nil {
+		return nil, err
+	}
+	ffile, err := basedir.Open(fmt.Sprintf("free_%03d", index))
+	if err != nil {
+		return nil, err
+	}
+	sl := newSlots(ffile.(sharkyFile), s.wg)
+	err = sl.load()
+	if err != nil {
+		return nil, err
+	}
+	sh := &shard{
+		reads:    make(chan read),
+		errc:     make(chan error),
+		writes:   s.writes,
+		index:    index,
+		datasize: datasize,
+		file:     file.(sharkyFile),
+		slots:    sl,
+		quit:     s.quit,
+	}
+	terminated := make(chan struct{})
+	sh.slots.wg.Add(1)
+	go func() {
+		sh.process()
+		close(terminated)
+	}()
+	go sl.process(terminated)
+	return sh, nil
+}
+
+// Read reads the content of the blob found at location into the byte buffer given
+// The location is assumed to be obtained by an earlier Write call storing the blob
+func (s *Store) Read(ctx context.Context, loc Location, buf []byte) (err error) {
+	sh := s.shards[loc.Shard]
+	select {
+	case sh.reads <- read{buf[:loc.Length], loc.Slot, 0}:
+		s.metrics.TotalReadCalls.Inc()
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	select {
+	case err = <-sh.errc:
+		if err != nil {
+			s.metrics.TotalReadCallsErr.Inc()
+		}
+		return err
+	case <-s.quit:
+		return ErrQuitting
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Write stores a new blob and returns its location to be used as a reference
+// It can be given to a Read call to return the stored blob.
+func (s *Store) Write(ctx context.Context, data []byte) (loc Location, err error) {
+	if len(data) > s.datasize {
+		return loc, ErrTooLong
+	}
+	s.wg.Add(1)
+	defer s.wg.Done()
+
+	c := s.pool.Get().(chan entry)
+	defer s.pool.Put(c)
+
+	select {
+	case s.writes <- write{data, c}:
+		s.metrics.TotalWriteCalls.Inc()
+	case <-s.quit:
+		return loc, ErrQuitting
+	case <-ctx.Done():
+		return loc, ctx.Err()
+	}
+
+	select {
+	case e := <-c:
+		if e.err == nil {
+			shard := strconv.Itoa(int(e.loc.Shard))
+			s.metrics.CurrentShardSize.WithLabelValues(shard).Add(1)
+		} else {
+			s.metrics.TotalWriteCallsErr.Inc()
+		}
+		return e.loc, e.err
+	case <-s.quit:
```

Comment:
there still a possible race condition here... if you quit or context is cancelled, the operation may still be in progress but the operation item may be put back into the sync pool, causing another operation to gobble up an item that is still actively used by another operation. the other operation would then read into the buffer while another operation is writing/reading from it. i would remove the `sync.Pool` altogether at this stage as it is just causing pain to manage (or just use it in cases we can reason about - this is not one of them)